### PR TITLE
😭 Fix use field array

### PR DESCRIPTION
**Pull request by bluebill1049**
Sorry guys for not produce high quality custom hook with `useFieldArray`. I am going to take my time over the weekend to polish those issues around `useFieldArray`. If you are using it, i would recommend for this PR to be merged and published. Thanks for your patient, and feel free to help me and us on testing. (🙏🏻 Building field array uncontrolled is completely new to me as well, hence made few mistakes here and there)

- [x] remove mapID of each render
  - for native input you will have to use `register({})` to trigger ref callback
- [x] fix errors got removed after remove item
- [x] fix delete issue
- [x] dirty
- [x] async reset dirty

### CODE REVIEW

**Comment by bluebill1049** [comment]
Early beta version: 4.6.3-beta.2 (Please help testing) ❤️

File: app/src/useFieldArray.tsx [hunk 1]
```diff
@@ -44,7 +44,7 @@ const UseFieldArray: React.FC = (props: any) => {
                 name={`data[${index}].name`}
                 defaultValue={data.name}
                 data-order={index}
-                ref={register}
+                ref={register({})}
```
Comment (kotarella1110) [inline app/src/useFieldArray.tsx]: Why do you need an empty object?

File: src/logic/getFieldValueByName.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import transformToNestObject from './transformToNestObject';
+import getFieldsValues from './getFieldsValues';
+
+export default <T, K extends keyof T>(fields: T, name?: K) => {
+  const results = transformToNestObject(getFieldsValues(fields));
+  return name ? results[name] : results;
```
Comment (kotarella1110) [inline src/logic/getFieldValueByName.ts]: dot syntax not supported for name.

**Review by kotarella1110 — COMMENTED** [review]
@bluebill1049 Can I commit to this branch to support dot syntax for name?

File: app/src/useFieldArray.tsx (hunk 1 above)
Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: for native input you will have to use register({}) to trigger ref callback during map.

```
eg: ref={(e) => { console.log('e') }}
```
will alway print the ref

```
const test= (e) => console.log(e);
ref={test}
```

will print the ref, and unmount will return null

File: src/logic/getFieldValueByName.ts (hunk 2 above)
Comment (bluebill1049) [reply src/logic/getFieldValueByName.ts]: i guess it's safer to do `get(results, name)`?

File: app/src/useFieldArray.tsx (hunk 1 above)
Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: also i will update the doc for this, and this only apply to `useFieldArray`

File: src/logic/getFieldValueByName.ts (hunk 2 above)
Comment (kotarella1110) [reply src/logic/getFieldValueByName.ts]: yes! that's what I think.

File: app/src/useFieldArray.tsx (hunk 1 above)
Comment (kotarella1110) [reply app/src/useFieldArray.tsx]: I'll look into this!

File: src/logic/getFieldValueByName.test.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  },
+  'nested.test[2].what[1].test': {
+    ref: {
+      name: 'nested.test[2].what[1].test',
+      value: 'nested.test[2].what[1].test',
+    },
+  },
+};
+
+describe('getFieldValueByName', () => {
```
Comment (bluebill1049) [inline src/logic/getFieldValueByName.test.ts]: ❤️

**Comment by bluebill1049** [comment]
latest beta: 4.7.0-beta.1

File: app/src/useFieldArray.tsx (hunk 1 above)
Comment (kotarella1110) [reply app/src/useFieldArray.tsx]: @bluebill1049 What is the bug caused by `ref={register}`?

Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: each time you remove, prepend, swap, move requires re-render, hence each ref needs to be updated. 

do an example below:
```
const test = (e) => console.log(e);
{[1,2,3].slice(0, index).map(() => <input ref={test} />)}
```
make a button to remove array index at a time, check the log

Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: ideally if we can support `register()` is better than asking people to put `register({})`

Comment (kotarella1110) [reply app/src/useFieldArray.tsx]: This problem can be solved by removing `useCallback`.

https://github.com/react-hook-form/react-hook-form/blob/5494803a59a1432ab2dd6ccbd6d574c95da833b0/src/useForm.ts#L1106

Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: yea good point, if we remove useCallback, then during useEffect user will get infinite loop

Comment (kotarella1110) [reply app/src/useFieldArray.tsx]: @bluebill1049 Sorry. Can you give an example where an infinite loop is triggered?
The following is an example of removing useCallback.

https://codesandbox.io/s/react-hook-form-usefieldarray-612i2

Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: try this: 

```
useEffect(() => {
 register(xxx)
}, [register])
```

Comment (kotarella1110) [reply app/src/useFieldArray.tsx]: Tried but no infinite loop :cry:

Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: really? even with useCallback?? wow

Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: it should run into infinite loop without useCallback with register

Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: FYI, it will lead to an issue like this one: https://github.com/react-hook-form/react-hook-form/issues/899

Comment (stramel) [reply app/src/useFieldArray.tsx]: @bluebill1049 My developer just ran into this. Do we have an issue to track and fix this?

Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: @stramel did you try to use just `register()` instead of just `register`

Comment (stramel) [reply app/src/useFieldArray.tsx]: Yeah, that's a fine workaround but is very unintuitive from a DX perspective

Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: yea true that, i have spend quite bit time on it. The problem is how ref is gets invoked when item mount and unmount. 

The following shows the diff

```
ref={() => {
  console.log('always called');
}}

const test = () => {
    console.log('only mount and unmount');
}

ref={test}
```

I haven't found a solution for it jus yet, that's why documented on the website.

Comment (stramel) [reply app/src/useFieldArray.tsx]: Can we create an issue to track this? I might take a look into it further

Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: Yes sure 👍

**Pull request merged**